File: 49-Group-Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
    const result = new Map();

    for (const word of strs) {
        const freq = new Array(26).fill(0);
        for (let i = 0; i < word.length; ++i) {
            freq[word[i].charCodeAt(0) - 'a'.charCodeAt(0)]++;
        }

        const key = freq.join('#');
        if (!result.has(key)) {
            result.set(key, []);
        }

        result.get(key).push(word);
    }
    return [...result.values()];
};

File: 49-Group-Anagrams.cpp
#define IOS ios_base::sync_with_stdio(false), cin.tie(NULL), cout.tie(0);
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        IOS 
        unordered_map<string, vector<string>> mp;

        for (auto& str : strs) {
            string word = str;
            sort(word.begin(), word.end());
            mp[word].push_back(str);
        }

        vector<vector<string>> result;
        for (auto& kv : mp)
            result.push_back(kv.second);

        return result;
    }
};